### ISSUE TITLE
Possible to prevent multiple instances of bot from running at the same time?

### ISSUE THREAD

**Issue by casey** [serenity-rs/serenity#1054]
Is it possible to prevent multiple instances of a bot from running at the same time?

I looked through the docs, but wasn't able to figure out how.

**Comment by arqunis** [comment]
Could you explain what you mean by "running at the same time"? Do you mean two `Client`s instances in the same process, or two `Client`s in different processes? If it's the latter, there's nothing that the library can do about that. Discord allows many connections to the same account, so this is legal. If it's the former, the only way for that to occur is if you `start`ed two `Client`s in different threads, but other than that, the library still has no mechanisms to prevent this. In fact, the library is designed to allow multiple `Client`s to run in parallel in the same process.

If you what you want to avoid is the latter situation, you could employ a file lock. Before you create a `Client`, you check for the existence of a file. If it doesn't exist, you create one and start the client, and upon shutting down your bot, you remove the file. If it does exist, you exit out of the program. This will ensure only one `Client` in a single process can be run.

**Comment by casey** [comment]
I'd like to prevent multiple instances of a bot binary from connecting to the same account, even if they're started on different machines. A file lock wouldn't work, because I might have the binary running on two different machines, or under two different accounts.

This is desirable because if I accidentally start two instances, the bot responds to actions twice. Ideally, everything the bot does will be idempotent, but I'd like to avoid making that a hard requirement.

Is there some way of getting the number of currently connected clients to the account? If I could do that, then perhaps I could implement this myself.

**Comment by vicky5124** [comment]
https://serenity-rs.github.io/serenity/current/serenity/model/gateway/struct.SessionStartLimit.html
You could probably be clever with this, and close the session if the number of allowed sessions is below the max-1

**Comment by casey** [comment]
@nitsuga5124 That's a good thought! Although, if the ratelimit period is 10 minutes (I'm not sure, but that's what I read in the API docs) then that might not work if an instance is launched, killed, and then relaunched within a 10 minute window.

**Comment by vicky5124** [comment]
What would be ideal is for serenity to support the gateway return value that tells you how many connections you have right now; i have seen it on other libraries such as python's hikari, that log how many sessions you have open of the bot.

**Comment by casey** [comment]
That would be awesome. Is that return value mentioned in the docs? I couldn't find it in the gateway docs.

**Comment by kangalio** [comment]
> i have seen it on other libraries such as python's hikari, that log how many sessions you have open of the bot.

@vicky5124 Where in hikari have you seen that? I can't find it (neither in the official Discord docs)

**Comment by vicky5124** [comment]
https://github.com/hikari-py/hikari/blob/b33a5c7b392d9de0afdeb271940c81865aaf252f/hikari/impl/bot.py#L1040 @kangalioo

**Comment by kangalio** [comment]
I see. But that count of open sessions resets after `reset_after` milliseconds. In [Discord's example payload](https://discord.com/developers/docs/topics/gateway#session-start-limit-object-session-start-limit-structure), that's 14400000 which is 4 hours. So, this approach can only detect multiple sessions if they were started within 4 hours.

I'm afraid it is impossible to do what OP is asking for (well; apart from the obvious solution of tracking open sessions yourself on some web service. But that's independent from the Discord API and serenity)

**Comment by vicky5124** [comment]
Tracking within 4 hours is still nice to have, usually when multiple sessions are an issue is when you are testing the bot, and previous sessions are left open by accident, not in production.

**Comment by kangalio** [comment]
Serenity already supports the endpoint in question ([link](https://docs.rs/serenity/latest/serenity/http/client/struct.Http.html#method.get_bot_gateway)) so I tested it. That tracker of sessions in the last 4 hours only tracks the number of sessions _started_, not currently _active_. So we're back to square one. I'll close this issue unless someone has a new idea how this feature request might be possible
